### DISCUSSION TITLE
Is there a Prisma way to run specific code inside a subscriber/hook like "after delete" or something like that?

### DISCUSSION THREAD

**Discussion by jelhouss** [prisma/prisma#21963] [Q&A]
### Question

Hey guys, I asked this question on Stack overflow but sadly I had no responses, and I am doing this manually right now and it sucks. I will write my issue and explain what I want to achieve.

I have a parent board that has many lists, each list has many tasks, each task can have many attachments, so it goes like this Board -> List -> Task -> Attachment.

Each attachment/file data (the attachment itself) is stored in an Object Storage server (MinIO for my case, could be S3 or whatever). Every time I delete a single attachment or all attachments I also clear my bucket in parallel as well, and for now I only run that MinIO deletion when I delete attachments from the task, e.g. `DELETE /task/:taskId/attachment` (delete all attachments under a task), or `DELETE /attachment/:attachmentId` (delete a specific attachment).

If I delete the list itself that is full of tasks, or even delete the board itself... it means that I need to perform that MinIO operation in all of those contexts, which is really bad, I don't wanna keep track of every delete operation that would require to clear the bucket.

Is there any Prisma-way to do this in a way that works like a subscriber or some kind of pubsub thing? TypeORM and MikroORM that treats tables as entities (and that repository pattern stuff) has a built-in subscribers that you can use to hook up with every operation like update or insert or delete of any kind of entity with metadata...

I read the Prisma docs, they offer [middlewares](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) but they're now deprecated as of the time I am writing this, and they offer another API called [extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/query), and nothing of these actually focus on this specific issue, they can help you to log what happened, or extend a built-in API like `findMany` or `create`, and this is not what I am looking for. Also there is this [$on](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#on) function and that also I am not sure it's suited for what I am looking for as it seems to be useful for logging (just displaying things) and not for performing some operations.

What I need is a way for me to run my minio code to clear the bucket with a given file name every time a file gets removed either by a direct request on the file resource or any of its parents.

I'm using Prisma 5.* (5.5.2) and Postgres. And I am also using Nest.js and this is not a specific issue for Nest.js I guess, I just want to figure out how to deal with automatic repetitive operation in a good Prisma way.

I tried looking for the docs and many GitHub issues, nothing found that could lead to something.

Help is really appreciated!

**Comment by ludralph** [comment]
Hi @jelhouss 👋 

Thank you for raising this question.

Middlewares are deprecated, which only means they will go in the future and that we might not invest into them as much as before.
It does not mean that no one can use them any more - they work just fine.

We also have a product called [Pulse](https://www.prisma.io/data-platform/pulse) that could potentially be used to address your use case. Pulse is a managed change data capture (CDC) service that captures change events from your database and delivers them instantly to your applications ([source](https://www.prisma.io/docs/data-platform/pulse/what-is-pulse)). With Pulse, you can subscribe to real-time data changes in your database, which could include delete operations on your Board, List, Task, or Attachment entities.
You can take a look at the [Getting Started](https://www.prisma.io/docs/getting-started/quickstart) guide to see how you begin using Pulse.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by jelhouss** [comment]
Hi @ludralph 

I am not sure middlewares will help. I didn't try to explore them deeply because on the surface they don't seem like the good place to put a code that depends on metadata like the instance type and id that is being affected either by a direct manipulation or by a parent manipulation in that tree of parent to parent/child to other parent/other child, etc... so I am not sure a middleware is capable of catching that and providing the related information I need.

Regarding Pulse, I have heard about and it seems that it still in early access process and I don't know how it's gonna be available yet either as a third-party piece of package or a built-in core or event a separate product, I don't know really what is gonna be like. Otherwise, it's not an option at the moment as far as I can tell, and also it feels overwhelming to relate to a CDC feature you know, I mean a CDC is more DB related compared to just an event that get triggered/handled by the ORM side just how some ORMs provides subscribers/hooks like in TypeORM or MikroORM for instance.

I will close the discussion now, but for anyone interested they can write a comment. Personally I used Nest.js interceptors and made all my MinIO calls to run right before I return my response and currently the interceptors seems to help organizing code, but the cons of it is that I will need to keep track of every resource I delete that I have to also delete my files from MinIO.

Thanks for your feedback
